src/canvasLogic/Canvas.ts

```diff
@@ -0,0 +1,175 @@
+        Math.floor(9 * Math.random()),
+    ].join("");
+
+    constructor() {}
+
+    public changeTool(tool: Tools_): void {
+        this.selectedTool = tool;
+    }
+
+    public changeShape(shape: Shapes_): void {
+        this.selectedShape = shape;
+    }
+
+    public down(event: IMouseEvent): void {
+        console.log(this.currentTool);
+        switch (this.selectedTool) {
+        case Tools.Shapes:
+            switch (this.selectedShape) {
+            case Shapes.Rectangle:
+                this.currentModifiedObject = new Rectangle(event.x, event.y);
+                break;
+            case Shapes.Ellipse:
+                this.currentModifiedObject = new Ellipse(event.x, event.y);
+                break;
+            case Shapes.Line:
+                this.currentModifiedObject = new Line(event.x, event.y);
+                break;
+            case Shapes.Arrow:
+                this.currentModifiedObject = new Arrow(event.x, event.y);
+                break;
+            default:
+                throw new IncorrectShapeError(this.selectedShape);
+            }
+            break;
+        case Tools.Text:
+            this.currentModifiedObject = new Text(event.x, event.y);
+            break;
+        default:
+            this.currentModifiedObject = null;
+            if (
+                this.selectedTool !== Tools.Cursor
+          && this.selectedTool !== Tools.Pen
+            ) {
+                throw new IncorrectToolError(this.selectedTool);
+            }
+        }
+    }
+
+    public move(event: IMouseEvent): void {
+        if (!this.currentModifiedObject) return;
+
+        switch (this.selectedTool) {
+        case Tools.Shapes:
+            if (!(this.currentModifiedObject instanceof Shape)) return;
+            this.currentModifiedObject.change(event.x, event.y);
+            break;
+        default:
+            this.currentModifiedObject = null;
+            if (
+                this.selectedTool !== Tools.Cursor
+          && this.selectedTool !== Tools.Text
+          && this.selectedTool !== Tools.Pen
+            ) {
+                throw new IncorrectToolError(this.selectedTool);
+            }
+        }
+    }
+
+    public up(event: IMouseEvent): void {
+        if (!this.currentModifiedObject) return;
+
+        switch (this.selectedTool) {
+        case Tools.Shapes:
+            if (!(this.currentModifiedObject instanceof Shape)) return;
+            this.currentModifiedObject.end(event.x, event.y);
+            break;
+        default:
+            this.currentModifiedObject = null;
+            if (
+                this.selectedTool !== Tools.Cursor
+          && this.selectedTool !== Tools.Text
+          && this.selectedTool !== Tools.Pen
+            ) {
+                throw new IncorrectToolError(this.selectedTool);
+            }
+        }
+    // TODO: Make it work (send an event to the parent)
+    // if ( this.selectedTool !== Tools.Pen ) this.changeTool(Tools.Cursor);
```

Comment:
Ouch. Don't let that slip into the main branch

Reply:
Well.... Could it be on dev branch? I really do need to do this but cannot resolve it now